### PR TITLE
Support hierarchical loading of the config file

### LINKED ISSUE #275: [Feature Request] Search ~/.ameba.yml as a default config for current linux user.
As done like rubocop.

we probably have to merge project config with home config.

Thank you.

### PR DESCRIPTION
**Pull request by Sija**
Files are searched for in the same order as used by [RuboCop](https://docs.rubocop.org/rubocop/configuration.html#config-file-locations).

Resolves #275

### CODE REVIEW

File: src/ameba/config.cr [hunk 1]
```diff
@@ -24,7 +24,12 @@ class Ameba::Config
     json:     Formatter::JSONFormatter,
   }
 
-  PATH = ".ameba.yml"
+  DEFAULT_PATHS = {
+    "~/.ameba.yml",
+    "~/.config/ameba/config.yml",
```
Comment (veelenga) [inline src/ameba/config.cr:29]: i think this must follow [XDG spec](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) and use the `${XDG_CONFIG_HOME:-~/.config}/ameba/` folder.

Comment (veelenga) [reply src/ameba/config.cr:29]: something like this?

```cr
XDG_CONFIG_HOME = ENV.fetch("XDG_CONFIG_HOME", "~/.config")

DEFAULT_PATHS = {
  "~/.ameba.yml",
  File.expand_path("#{XDG_CONFIG_HOME}/ameba", home: true)
}
```

Comment (Sija) [reply src/ameba/config.cr:29]: Yeah, no need for `.expand_path` though, since it's being already done l8r on

Comment (Sija) [reply src/ameba/config.cr]: Fixed, thanks for spotting! 🙏

**Review by veelenga — APPROVED** [review]
LGTM

**Comment by Sija** [comment]
@veelenga Should we create v1.4.3 (I guess the last in 1.4.x branch) to release it? WDYT?

File: src/ameba/config.cr [hunk 2]
```diff
@@ -77,14 +84,33 @@ class Ameba::Config
...
 
+  protected def self.read_config(path) : String?
+    if path
+      return File.exists?(path) ? File.read(path) : nil
+    end
+    path = Path[DEFAULT_PATH].expand(home: true)
+
+    search_paths = path
+      .parents
+      .map! { |search_path| search_path / FILENAME }
```
Comment (straight-shoota) [inline src/ameba/config.cr:103]: Why not do this lazily while iterating?

**Review by straight-shoota — COMMENTED** [review]
A more DRY approach for `read_config` would be something like this:

```cr
def read_config(file)
  each_config_path(file) do |path|
    if File.exists?(path)
      return File.read(path)
    end
  end
end

def each_config_path(path)
  yield path if path
  # ...
  search_paths.reverse_each do |search_path|
    yield search_path
  end
  # ...
  DEFAULT_PATHS.each do |default_path|
      yield default_path
    end
end
```

**Comment by Sija** [comment]
@straight-shoota Looks good, I'll do that, thanks!

**Review by veelenga — APPROVED** [review]

**Comment by straight-shoota** [comment]
If you want completely different behaviour depending on whether `path` is present, I'd make that more explicit. Probably with two separate overloads, `read_config(String | Path)` and `read_config(Nil)` (or `read_config()`).

File: src/ameba/config.cr [hunk 3]
```diff
@@ -77,14 +106,36 @@ class Ameba::Config
...
-    content = File.exists?(path) ? File.read path : "{}"
+    content = read_config(path) || "{}"
     Config.new YAML.parse(content)
   rescue e
     raise "Config file is invalid: #{e.message}"
   end
 
+  protected def self.read_config(path = nil)
+    if path
+      return File.exists?(path) ? File.read(path) : nil
```
[the remark this answers is not in the document]
Comment (veelenga) [reply src/ameba/config.cr:119]: Yes, I think it should.

[the remark this answers is not in the document]
Comment (Sija) [reply src/ameba/config.cr:119]: I'll create a new issue/pr to handle that.

**Pull request merged**